If a page changes its title while in the background, allow the page to run in the background

cdumez opened the pull request:
#### e1f40e7ef94dc4020e1efe7f83e2d360cee11d9b
<pre>
If a page changes its title while in the background, allow the page to run in the background
<a href="https://bugs.webkit.org/show_bug.cgi?id=256259">https://bugs.webkit.org/show_bug.cgi?id=256259</a>
rdar://103388697

Reviewed by Ben Nham.

If a page changes its title while in the background, allow the page to run in the
background, as long as the app is foreground. Previously, the page changing its
title would buy its an extra 8 minutes of background running time. We are now
more permissive and allow the page to run indefinitely in the background when we
detect this (as long as the application is foreground).

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::dropSuspendedAssertionTimerFired):
(WebKit::ProcessThrottler::pageAllowedToRunInTheBackgroundToken):
(WebKit::ProcessThrottler::numberOfPagesAllowedToRunInTheBackgroundChanged):
(WebKit::ProcessThrottler::delaySuspension): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didFailLoadForFrame):
(WebKit::WebPageProxy::didReceiveTitleForFrame):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
(WebKit::WebPageProxy::ProcessActivityState::takeLikelyToUseNotificationsActivity): Deleted.
(WebKit::WebPageProxy::ProcessActivityState::dropLikelyToUseNotificationsActivity): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
</pre>

bnham commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -336,11 +340,23 @@ void ProcessThrottler::setShouldTakeSuspendedAssertion(bool shouldTakeSuspendedA
 {
-    PROCESSTHROTTLER_RELEASE_LOG("delaySuspension");
+    if (m_pageAllowedToRunInTheBackgroundCounter.value())
+        return;
```
We delete the assertion when `m_pageAllowedToRunInTheBackgroundCounter` goes from 1 => 0 but we don't create one when it goes from 0 => 1. So I guess this RefCounter is only really meant to suppress dropping the assertion and `setThrottleState` is still in charge of creating an assertion if necessary?

cdumez replied:
The idea is that if you're able to go from 0 to 1, then you were not suspended. If you weren't suspended then we were holding a process assertion already.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5585,9 +5563,7 @@ void WebPageProxy::didCommitLoadForFrame(FrameIdentifier frameID, FrameInfoData&
-#if PLATFORM(MAC)
-        m_processActivityState.dropLikelyToUseNotificationsActivity();
-#endif
+        internals().pageAllowedToRunInTheBackgroundToken = nullptr;
```
Should we drop this in `didFailLoadForFrame` as well?

cdumez replied:
Yes, we probably should. I'll fix.

cdumez committed:
If a page changes its title while in the background, allow the page to run in the background
https://bugs.webkit.org/show_bug.cgi?id=256259
rdar://103388697

Reviewed by Ben Nham.

If a page changes its title while in the background, allow the page to run in the
background, as long as the app is foreground. Previously, the page changing its
title would buy its an extra 8 minutes of background running time. We are now
more permissive and allow the page to run indefinitely in the background when we
detect this (as long as the application is foreground).

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::dropSuspendedAssertionTimerFired):
(WebKit::ProcessThrottler::pageAllowedToRunInTheBackgroundToken):
(WebKit::ProcessThrottler::numberOfPagesAllowedToRunInTheBackgroundChanged):
(WebKit::ProcessThrottler::delaySuspension): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didReceiveTitleForFrame):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
(WebKit::WebPageProxy::ProcessActivityState::takeLikelyToUseNotificationsActivity): Deleted.
(WebKit::WebPageProxy::ProcessActivityState::dropLikelyToUseNotificationsActivity): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: https://commits.webkit.org/263651@main
